crates/e2e-testing/src/spin.rs

```diff
@@ -0,0 +1,88 @@
+use crate::utils;
+use anyhow::Result;
+use std::path::{Path, PathBuf};
+use std::process::Output;
+use std::sync::Mutex;
+
+#[cfg(target_family = "unix")]
+use {
+    nix::sys::signal::{kill, Signal},
+    nix::unistd::Pid,
+};
+
+static INSTALLING_TEMPLATES_MUTEX: Mutex<i32> = Mutex::new(0);
+static INSTALLING_PLUGINS_MUTEX: Mutex<i32> = Mutex::new(0);
+
+pub fn template_install(mut args: Vec<&str>) -> Result<Output> {
+    let mut cmd = vec!["spin", "templates", "install"];
+    cmd.append(&mut args);
+
+    let mut x = INSTALLING_TEMPLATES_MUTEX.lock().unwrap();
+    let result = utils::run(cmd, None, None);
+
+    //this ensure we have mutex lock until here
+    *x += 1;
+
+    result
+}
+
+pub fn new_app(template_name: &str, app_name: &str) -> Result<Output> {
+    let basedir: PathBuf = [env!("CARGO_MANIFEST_DIR"), "..", "..", "tests", "testcases"]
+        .iter()
+        .collect();
+
+    return utils::run(
+        vec!["spin", "new", template_name, app_name, "--accept-defaults"],
+        basedir.to_str(),
+        None,
+    );
+}
+
+pub fn install_plugins(plugins: Vec<&str>) -> Result<Output> {
+    // lock mutex to ensure one install_plugins runs at a time
+    let mut x = INSTALLING_PLUGINS_MUTEX.lock().unwrap();
+
+    let mut output = utils::run(vec!["spin", "plugin", "update"], None, None)?;
+
+    for plugin in plugins {
+        output = utils::run(
+            vec!["spin", "plugin", "install", plugin, "--yes"],
+            None,
+            None,
+        )?;
+    }
+
+    //this ensure we have mutex lock until here
+    *x += 1;
```

Comment:
Newb Rust question: I would have expected you to do something like `INSTALLING_PLUGINS_MUTEX.unlock().unwrap();` here rather than `*x += 1;`. Can you help me understand what `*x += 1;` is doing?

Reply:
i think calling `INSTALLING_PLUGINS_MUTEX.unlock().unwrap();` is like calling `mutex.Lock()`, and the last change done to the value retrieved via that call automatically unlocks the mutex.

incrementing the value of x is just a hack to tell rust to lock the mutex until this line.

but i am very new rust myself too, so if there is a better way to achieve this, i would be very glad to learn and implement that.

Reply:
Weirdly, it seems like `Mutex::unlock()` is nightly-only for now, so the question is moot!  https://doc.rust-lang.org/std/sync/struct.Mutex.html#method.unlock

For what it's worth, the idiomatic way of dropping the lock in Rust is to... `drop` the lock object (`x` in this case).  `x` is a `MutexGuard` and when it goes out of scope, its `Drop` implementation unlocks the mutex a la RAII (https://doc.rust-lang.org/std/sync/struct.Mutex.html#method.lock).  There is no need to modify the content.  If you're concerned about `x` being dropped before the end of the (lexical) scope, you can write an explicit `drop(x);` - that is probably clearer than doing what I think is an unused mutation to force `x` to remain in scope.

Reply:
Looking at the current code, the lock _should_ be dropped at the closing brace on line 27.  But if you want to ensure a lifetime against incorrect future maintenance, then `drop` will do the job.

Reply:
yeah, this looks so much better. replaced it to `drop(x)`;